lib/auth/version.go

```diff
@@ -20,55 +20,101 @@ package auth
-	currentVersion *semver.Version,
-) error {
+	procStorage VersionStorage,
+	backendStorage services.AuthInfoService,
+	currentVersion semver.Version,
+) (err error) {
 	if skip := os.Getenv(skipVersionUpgradeCheckEnv); skip != "" {
 		return nil
 	}
 
-	lastKnownVersion, err := storage.GetTeleportVersion(ctx)
-	if trace.IsNotFound(err) {
-		if err := storage.WriteTeleportVersion(ctx, currentVersion); err != nil {
+	for range versionUpgradeCheckMaxWriteRetry {
+		authInfo, err := backendStorage.GetAuthInfo(ctx)
+		if trace.IsNotFound(err) {
+			// TODO(vapopov): DELETE IN v19.0.0 – the last known version should already be migrated to backend storage.
+			// Fallback to local process storage for backward compatibility with previous versions.
+			lastKnownVersion, err := procStorage.GetTeleportVersion(ctx)
+			if trace.IsNotFound(err) {
+				authInfo, err = authinfo.NewAuthInfo(&authinfov1.AuthInfoSpec{TeleportVersion: currentVersion.String()})
+				if err != nil {
+					return trace.Wrap(err)
+				}
+				if _, err := backendStorage.CreateAuthInfo(ctx, authInfo); err != nil {
+					err = trace.Wrap(err)
+					slog.WarnContext(ctx, "Failed to create AuthInfo resource",
+						"error", err)
+					continue
+				}
+				return nil
+			} else if err != nil {
+				return trace.Wrap(err)
+			}
+
+			authInfo, err = authinfo.NewAuthInfo(&authinfov1.AuthInfoSpec{TeleportVersion: lastKnownVersion.String()})
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			if _, err := backendStorage.CreateAuthInfo(ctx, authInfo); err != nil {
+				err = trace.Wrap(err)
+				slog.WarnContext(ctx, "Failed to create AuthInfo resource",
+					"error", err)
+				continue
+			}
+		} else if err != nil {
 			return trace.Wrap(err)
 		}
-		return nil
-	} else if err != nil {
-		return trace.Wrap(err)
-	}
 
-	if currentVersion.Major-lastKnownVersion.Major > 1 {
-		return trace.BadParameter("Unsupported upgrade path detected: from %v to %v. "+
-			"Teleport supports direct upgrades to the next major version only.\n Please upgrade "+
-			"your cluster to version %d.x.x first. See compatibility guarantees for details: "+
-			"https://goteleport.com/docs/upgrading/overview/#component-compatibility.",
-			lastKnownVersion, currentVersion.String(), lastKnownVersion.Major+1)
-	}
-	if lastKnownVersion.Major-currentVersion.Major > 1 {
-		return trace.BadParameter("Unsupported downgrade path detected: from %v to %v. "+
-			"Teleport doesn't support major version downgrade.\n Please downgrade "+
-			"your cluster to version %d.x.x first. See compatibility guarantees for details: "+
-			"https://goteleport.com/docs/upgrading/overview/#component-compatibility.",
-			lastKnownVersion, currentVersion.String(), lastKnownVersion.Major-1)
-	}
-	if err := storage.WriteTeleportVersion(ctx, currentVersion); err != nil {
-		return trace.Wrap(err)
+		lastKnownVersion, err := semver.NewVersion(authInfo.GetSpec().GetTeleportVersion())
+		if err != nil {
+			return trace.Wrap(err)
```

Comment:
This error condition deserves a better error message and some steps that the cluster admin can take to fix the situation (broken backend item or invalid version string).